File: version_negotiator.py
"""Pure version negotiation between two AnyClip peers.

Decides whether two instances can link based on their app + protocol
versions. Kept side-effect free so it can be unit tested with the table
of cases in tests/test_version_negotiator.py and so the daemon code in
anyclip.py only needs to consume the resulting enum.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class Compatibility(Enum):
    COMPATIBLE = "compatible"
    PEER_OLDER_MINOR = "peer_older_minor"
    PEER_NEWER_MINOR = "peer_newer_minor"
    PEER_OLDER_MAJOR = "peer_older_major"
    PEER_NEWER_MAJOR = "peer_newer_major"


@dataclass(frozen=True)
class VersionInfo:
    app_version: str
    protocol_major: int
    protocol_minor: int


def negotiate(local: VersionInfo, peer: VersionInfo) -> Compatibility:
    """Compare local vs peer protocol versions.

    Major version dominates: any major mismatch is a refusal regardless
    of minor. Minor differences are advisory and keep the link.
    ``app_version`` is informational and never affects the outcome.
    """
    if peer.protocol_major < local.protocol_major:
        return Compatibility.PEER_OLDER_MAJOR
    if peer.protocol_major > local.protocol_major:
        return Compatibility.PEER_NEWER_MAJOR
    if peer.protocol_minor < local.protocol_minor:
        return Compatibility.PEER_OLDER_MINOR
    if peer.protocol_minor > local.protocol_minor:
        return Compatibility.PEER_NEWER_MINOR
    return Compatibility.COMPATIBLE


def link_allowed(result: Compatibility) -> bool:
    """True if the link should be kept open for this compatibility result."""
    return result in {
        Compatibility.COMPATIBLE,
        Compatibility.PEER_OLDER_MINOR,
        Compatibility.PEER_NEWER_MINOR,
    }

File: formacOS/Sources/AnyClipCore/VersionNegotiator.swift
/// Pure version negotiation between two AnyClip peers.
/// Port of version_negotiator.py — keep the table in lockstep.

public enum Compatibility: String, Sendable, Equatable {
    case compatible = "compatible"
    case peerOlderMinor = "peer_older_minor"
    case peerNewerMinor = "peer_newer_minor"
    case peerOlderMajor = "peer_older_major"
    case peerNewerMajor = "peer_newer_major"
}

public struct VersionInfo: Sendable, Equatable {
    public let appVersion: String
    public let protocolMajor: Int
    public let protocolMinor: Int

    public init(appVersion: String, protocolMajor: Int, protocolMinor: Int) {
        self.appVersion = appVersion
        self.protocolMajor = protocolMajor
        self.protocolMinor = protocolMinor
    }
}

/// Major version dominates: any major mismatch is a refusal regardless of
/// minor. Minor differences are advisory and keep the link. appVersion is
/// informational and never affects the outcome.
public func negotiate(local: VersionInfo, peer: VersionInfo) -> Compatibility {
    if peer.protocolMajor < local.protocolMajor { return .peerOlderMajor }
    if peer.protocolMajor > local.protocolMajor { return .peerNewerMajor }
    if peer.protocolMinor < local.protocolMinor { return .peerOlderMinor }
    if peer.protocolMinor > local.protocolMinor { return .peerNewerMinor }
    return .compatible
}

public func linkAllowed(_ result: Compatibility) -> Bool {
    switch result {
    case .compatible, .peerOlderMinor, .peerNewerMinor: return true
    case .peerOlderMajor, .peerNewerMajor: return false
    }
}
